onnxruntime/core/providers/rocm/tunable/gemm_hipblaslt.h

```diff
@@ -0,0 +1,207 @@
+  }
+  if (enable_scaleD) {
+    HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatmulDescSetAttribute(
+        matmul, HIPBLASLT_MATMUL_DESC_D_SCALE_POINTER, &d_scaleD, sizeof(void*)));
+  }
+
+  hipblasLtMatmulPreference_t pref;
+  void* workspace;
+  size_t max_workspace_size = 32 * 1024 * 1024;
+  HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatmulPreferenceCreate(&pref));
+  HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatmulPreferenceSetAttribute(
+      pref, HIPBLASLT_MATMUL_PREF_MAX_WORKSPACE_BYTES, &max_workspace_size, sizeof(max_workspace_size)));
+
+  const int heuristic_result_count = 3;
+  hipblasLtMatmulHeuristicResult_t heuristic_result[heuristic_result_count] = {0};
+  int ret_algo_count = 0;
+  HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatmulAlgoGetHeuristic(handle,
+                                                            matmul,
+                                                            mat_a,
+                                                            mat_b,
+                                                            mat_c,
+                                                            mat_c,
+                                                            pref,
+                                                            heuristic_result_count,
+                                                            heuristic_result,
+                                                            &ret_algo_count));
+
+  size_t workspace_size = heuristic_result[0].workspaceSize;
+  if (workspace_size > 0) {
+    HIP_RETURN_IF_ERROR(hipMallocAsync(&workspace, workspace_size, params->stream));
+  }
+
+  HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatmul(handle,
+                                            matmul,
+                                            &alpha,
+                                            params->b,
+                                            mat_a,
+                                            params->a,
+                                            mat_b,
+                                            &beta,
+                                            params->c,
+                                            mat_c,
+                                            params->c,
+                                            mat_c,
+                                            &heuristic_result[0].algo,
+                                            workspace,
+                                            workspace_size,
+                                            params->stream));
+
+  if (workspace_size > 0) {
+    HIP_RETURN_IF_ERROR(hipFreeAsync(workspace, params->stream));
+  }
+
+  HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatmulPreferenceDestroy(pref));
+  HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatmulDescDestroy(matmul));
+  HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatrixLayoutDestroy(mat_a));
+  HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatrixLayoutDestroy(mat_b));
+  HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatrixLayoutDestroy(mat_c));
+  HIPBLASLT_RETURN_IF_ERROR(hipblasLtDestroy(handle));
```

Comment:
>   HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatmulPreferenceDestroy(pref));
  HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatmulDescDestroy(matmul));
  HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatrixLayoutDestroy(mat_a));
  HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatrixLayoutDestroy(mat_b));
  HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatrixLayoutDestroy(mat_c));
  HIPBLASLT_RETURN_IF_ERROR(hipblasLtDestroy(handle));
  
  General question: are any of these structures expensive to initialize/create? Should we consider reusing them between calls?

Reply:
No they are not, they are generally small structs of ints as defined in [hipBLASLt handle.h](https://github.com/ROCmSoftwarePlatform/hipBLASLt/blob/develop/library/src/amd_detail/rocblaslt/src/include/handle.h)